# Recursion & Backtracking

## Topics Covered
- Fibonacci using Recursion
- Factorial using Recursion
- Tower of Hanoi
- LeetCode 509 - Fibonacci Number
- LeetCode 344 - Reverse String

---

Recursion = A function calling itself until a base condition is met.

General Structure:

1. Base Case
2. Recursive Call

Without a base case, recursion runs forever.


In [1]:
# Factorial using recursion

def factorial(n):
    if n == 0:
        return 1

    return n * factorial(n - 1)

print(factorial(5))

120


## Dry Run

factorial(5)

5 * factorial(4)

5 * 4 * factorial(3)

5 * 4 * 3 * factorial(2)

5 * 4 * 3 * 2 * factorial(1)

5 * 4 * 3 * 2 * 1

Answer = 120

In [3]:
# Fibonacci using recursion

def fibonacci(n):
    if n <= 1:
        return n

    return fibonacci(n - 1) + fibonacci(n - 2)

for i in range(10):
    print(fibonacci(i), end=" ")

0 1 1 2 3 5 8 13 21 34 

## Fibonacci Sequence

0 1 1 2 3 5 8 13 21 34

Formula:

fib(n) = fib(n-1) + fib(n-2)

Base Cases:

fib(0) = 0
fib(1) = 1

In [4]:
# Print numbers from n to 1 using recursion

def countdown(n):
    if n == 0:
        return

    print(n)
    countdown(n - 1)

countdown(5)

5
4
3
2
1


In [5]:
# Print numbers from 1 to n using recursion

def countup(n):
    if n == 0:
        return

    countup(n - 1)
    print(n)

countup(5)

1
2
3
4
5


## Tower of Hanoi

Rules:

1. Move one disk at a time
2. Bigger disk cannot be placed on smaller disk
3. Use auxiliary rod

Recursive Formula:

Move n-1 disks

Move largest disk

Move n-1 disks again

In [6]:
def tower_of_hanoi(n, source, target, auxiliary):
    if n == 1:
        print(f"Move disk 1 from {source} to {target}")
        return

    tower_of_hanoi(n - 1, source, auxiliary, target)

    print(f"Move disk {n} from {source} to {target}")

    tower_of_hanoi(n - 1, auxiliary, target, source)

tower_of_hanoi(3, "A", "C", "B")

Move disk 1 from A to C
Move disk 2 from A to B
Move disk 1 from C to B
Move disk 3 from A to C
Move disk 1 from B to A
Move disk 2 from B to C
Move disk 1 from A to C


In [7]:
# LeetCode 509
# Fibonacci Number

class Solution:
    def fib(self, n: int) -> int:
        if n <= 1:
            return n

        return self.fib(n - 1) + self.fib(n - 2)

In [8]:
# Optimized Fibonacci

class Solution:
    def fib(self, n: int) -> int:
        if n <= 1:
            return n

        a = 0
        b = 1

        for _ in range(2, n + 1):
            a, b = b, a + b

        return b

In [10]:
# LeetCode 344
# Reverse String

class Solution:
    def reverseString(self, s):
        def helper(left, right):
            if left >= right:
                return

            s[left], s[right] = s[right], s[left]

            helper(left + 1, right - 1)

        helper(0, len(s) - 1)